File: snapi-frontend/src/main/java/antlr4_parser/Antlr4SyntaxAnalyzer.java
/*
 * Copyright 2023 RAW Labs S.A.
 *
 * Use of this software is governed by the Business Source License
 * included in the file licenses/BSL.txt.
 *
 * As of the Change Date specified in that file, in accordance with
 * the Business Source License, use of this software will be governed
 * by the Apache License, Version 2.0, included in the file
 * licenses/APL.txt.
 */

package antlr4_parser;

import antlr4_parser.generated.SnapiLexer;
import antlr4_parser.generated.SnapiParser;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.bitbucket.inkytonik.kiama.parsing.Parsers;
import org.bitbucket.inkytonik.kiama.util.Position;
import org.bitbucket.inkytonik.kiama.util.Positions;
import org.bitbucket.inkytonik.kiama.util.StringSource;
import raw.compiler.base.source.BaseProgram;
import raw.compiler.base.source.Type;
import raw.compiler.rql2.source.Rql2Program;
import scala.Tuple2;
import scala.util.Either;
import scala.util.Right;

public class Antlr4SyntaxAnalyzer extends Parsers {

  private final Positions positions;

  public Antlr4SyntaxAnalyzer(Positions positions) {
    super(positions);
    this.positions = positions;
  }

  public Either<Tuple2<String, Position>, BaseProgram> parse(String s) {
    SnapiLexer lexer = new SnapiLexer(CharStreams.fromString(s));
    SnapiParser parser = new SnapiParser(new CommonTokenStream(lexer));
    ParseTree tree = parser.prog();

    System.out.println(tree.toStringTree(parser));

    RawSnapiVisitor visitor = new RawSnapiVisitor(new StringSource(s, ""), positions);
    Rql2Program result = (Rql2Program) visitor.visit(tree);
    return new Right<>(result);
  }

  public Either<Tuple2<String, Position>, Type> parseType(String s) {
    SnapiLexer lexer = new SnapiLexer(CharStreams.fromString(s));
    SnapiParser parser = new SnapiParser(new CommonTokenStream(lexer));
    ParseTree tree = parser.type();
    RawSnapiVisitor visitor = new RawSnapiVisitor(new StringSource(s, ""), positions);
    Type result = (Type) visitor.visit(tree);
    return new Right<>(result);
  }

  public Positions getPositions() {
    return positions;
  }
}

File: compiler-snapi-frontend/src/main/scala/com/rawlabs/compiler/snapi/rql2/antlr4/Antlr4SyntaxAnalyzer.scala
/*
 * Copyright 2024 RAW Labs S.A.
 *
 * Use of this software is governed by the Business Source License
 * included in the file licenses/BSL.txt.
 *
 * As of the Change Date specified in that file, in accordance with
 * the Business Source License, use of this software will be governed
 * by the Apache License, Version 2.0, included in the file
 * licenses/APL.txt.
 */

package com.rawlabs.compiler.snapi.rql2.antlr4

import com.rawlabs.compiler.api.Message
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.{CharStreams, CommonTokenStream, TokenStream}
import org.bitbucket.inkytonik.kiama.parsing.Parsers
import org.bitbucket.inkytonik.kiama.util.{Positions, Source, StringSource}
import com.rawlabs.compiler.snapi.base.source.Type
import com.rawlabs.compiler.snapi.common.source.SourceProgram
import com.rawlabs.compiler.snapi.generated.{SnapiLexer, SnapiParser}
import com.rawlabs.compiler.snapi.rql2.source.Rql2Program

abstract class ParseResult(errors: List[Message]) {
  def hasErrors: Boolean = errors.nonEmpty

  def isSuccess: Boolean = errors.isEmpty
}
final case class ParseProgramResult[P](errors: List[Message], tree: P) extends ParseResult(errors)
final case class ParseTypeResult(errors: List[Message], tipe: Type) extends ParseResult(errors)

class Antlr4SyntaxAnalyzer(val positions: Positions, isFrontend: Boolean) extends Parsers(positions) {

  def parse(s: String): ParseProgramResult[SourceProgram] = {
    val rawErrorListener = new RawErrorListener()
    val stream = getTokenStream(s, rawErrorListener)
    parse(stream, StringSource(s), rawErrorListener)
  }

  protected def parse(
      stream: TokenStream,
      source: Source,
      errorListener: RawErrorListener
  ): ParseProgramResult[SourceProgram] = {
    val parser = new SnapiParser(stream)
    parser.removeErrorListeners()
    parser.addErrorListener(errorListener)

    val tree: ParseTree = parser.prog
    val visitorParseErrors = RawVisitorParseErrors()
    val visitor = new RawSnapiVisitor(positions, source, isFrontend, visitorParseErrors)
    val result = visitor.visit(tree).asInstanceOf[Rql2Program]

    val totalErrors = errorListener.getErrors ++ visitorParseErrors.getErrors
    ParseProgramResult(totalErrors, result)
  }

  protected def getTokenStream(s: String, errorListener: RawErrorListener): CommonTokenStream = {
    val lexer = new SnapiLexer(CharStreams.fromString(s))
    lexer.removeErrorListeners()
    lexer.addErrorListener(errorListener)

    new CommonTokenStream(lexer)
  }

  def parseType(s: String): ParseTypeResult = {
    val source = StringSource(s)
    val rawErrorListener = new RawErrorListener()

    val parser = new SnapiParser(getTokenStream(s, rawErrorListener))
    parser.removeErrorListeners()
    parser.addErrorListener(rawErrorListener)

    val tree: ParseTree = parser.tipe
    val visitorParseErrors = RawVisitorParseErrors()
    val visitor: RawSnapiVisitor = new RawSnapiVisitor(positions, source, isFrontend, visitorParseErrors)
    val result: Type = visitor.visit(tree).asInstanceOf[Type]

    val totalErrors = rawErrorListener.getErrors ++ visitorParseErrors.getErrors

    ParseTypeResult(totalErrors, result)
  }

}
